### PR TITLE
Add integration tests that excercise Thanos Receive edge cases

### PR DESCRIPTION
**Pull request by ianbillett**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

This PR adds 3 new integration test scenarios for Thanos Receive.

`router_replication` and `router_tree` both currently fail against `main`.

See https://github.com/thanos-io/thanos/issues/4359 for discussion.

## Verification

Tests are failing :upside_down_face:

### CODE REVIEW

**Comment by GiedriusS** [comment]
Thanks for this! I am thinking that maybe let's convert `testutil.Ok` to `testutil.NotOk`, merge, and then fix this in a follow-up PR?

**Comment by ianbillett** [comment]
Nice idea @GiedriusS :+1: 

I've updated the integration tests to fail on the 'correct' value. PTAL.

**Comment by bwplotka** [comment]
`Add failing test` title sounds... scary ;p

File: test/e2e/receive_test.go [hunk 1]
```diff
@@ -101,51 +187,161 @@ func TestReceive(t *testing.T) {
...
+					   └───────┘  │   │  └───────┘
+							   ┌──▼───▼──┐
+							   │         │
+							   │ Router1 │
+						  ┌────┤         ├───────┐
+						  │    └─────────┘       │
+					  ┌───▼─────┐          ┌─────▼─────┐
+					  │         │          │           │
+					  │ Router2 │          │ Ingestor1 │
+				  ┌───┤         ├───┐      │           │
```
Comment (bwplotka) [inline test/e2e/receive_test.go:223]: Something looks off here

File: test/e2e/receive_test.go [hunk 2]
```diff
@@ -101,51 +187,161 @@ func TestReceive(t *testing.T) {
...
-				"prometheus": "prom2",
-				"receive":    "i1",
-				"replica":    "0",
-				"tenant_id":  "default-tenant",
+		}, 3)
+
+		// TODO(ianbillett): change this assertion to testutil.Equals when #4359 is fixed.
+		// Currently, we expect this e2e test to fail due to bugs in the reicever split proposal implementation.
+		for _, series := range result {
+			testutil.NotEqual(t, series.Value, 2.0)
```
Comment (bwplotka) [inline test/e2e/receive_test.go:197]: Why not equal something else with TODO?

File: pkg/testutil/testutil.go [hunk 3]
```diff
@@ -79,6 +79,21 @@ func Equals(tb testing.TB, exp, act interface{}, v ...interface{}) {
 	tb.Fatal(sprintfWithLimit("\033[31m%s:%d:"+msg+"\n\n\texp: %#v\n\n\tgot: %#v%s\033[39m\n\n", filepath.Base(file), line, exp, act, diff(exp, act)))
 }
 
+// NotEqual fails the test if exp is equal to act.
+func NotEqual(tb testing.TB, exp, act interface{}, v ...interface{}) {
```
Comment (bwplotka) [inline pkg/testutil/testutil.go:83]: I would be strict and suggest removing it. It would be much better to assert on the expected thing, then assert that thing is not value X ...  WDYT? (:

**Review by bwplotka — CHANGES_REQUESTED** [review]
Nice tests, thanks for those,

But I would not add new helper, see comment 🙈

File: test/e2e/receive_test.go (hunk 2 above)
Comment (bwplotka) [reply test/e2e/receive_test.go:197]: Essentially this can be surprising. So easy to match this assertion (so many options that passes). It essentially does not give any value

File: test/e2e/receive_test.go (hunk 1 above)
Comment (ianbillett) [reply test/e2e/receive_test.go:223]: Yeah - it renders properly in Goland :thinking: I'll have a play and see if I can make it look right.

File: pkg/testutil/testutil.go (hunk 3 above)
Comment (ianbillett) [reply pkg/testutil/testutil.go:83]: Yeah fair enough - adding this did feel a bit sketchy :sweat_smile:

**Review by bwplotka — APPROVED** [review]
I love it, thanks! 

Sorry for the massive lag on checking those!
